# RNAmap

`RNAmap` is a bash script that automates the processes of:
1. Trimming of adapter sequences and poor quality nucleotides (cutadapt)
2. Removing read pairs where one of the reads has less than 40bp
3. Quality Control (FastQC)
4. Mapping reads to the genome (STAR)
5. Read counting per gene (HTSeq)
6. Summary statistics 

The script currently works only if the following conditions are met
(possible to extend the script for other settings):
- Libraries were prepared using a *TruSeq stranded Kit*
- fastq data was generated on a MiSeq

## Example


```bash
script=/mnt/users/fabig/cluster_pipelines/RnaMapping/RNAmap.sh
fdir=/mnt/SeqData2/MiSeq_data/150612_M02210_0008_000000000-ACGHJ/Data/Intensities/BaseCalls
sdir=/mnt/SeqData2/MiSeq_data/150612_M02210_0008_000000000-ACGHJ/Data/Intensities/BaseCalls/SampleSheet.csv

bash $script -d $fdir -s $sdir -m SampleMaster.txt
```

### Reports

The script also submits 2 `R` scripts that generate both:

1. STAR mapping statistics (see: `star/STAR_Log_Stat.txt` and `star/STAR_Log_Stat.pdf`)
2. HTseq counting statistics (see: `count/HTSeq_Count_Stats.txt` and `count/HTSeq_Count_Stats.pdf`)

In order for the `R` scripts to work you need to have the following
packages installed in your `orion R`

- `ggplot2`
- `reshape`
- `gtable`

## RNAmap options

- `-d|--dirin`: Full path of the directory that holds the .fasq.gz
  files

- `-g|--genome`: *Optional*, defaults to the latest version of the _Salmon salar_ genome which is
located at:
`/mnt/users/fabig/RNAseq/Ssa_genome/CIG_3.6v2_chrom-NCBI/STAR_index`

- `-s|--samplesheet`:  *Optional*  Full path to the Illumina _sample
  sheet_ (.`.csv` format). 

- `-m|--mastersheet`: Name of the _master sheet_, this will be created
  in the current directory and contains all relevant sample
  information. Information is extracted from the _sample sheet_ and
  arranged in a more user friendly way. If argument `-s|--samplesheet`
  is empty then `-m|--mastersheet` needs to point to a _master sheet_
  file (see _master sheet_ format guide)

- `--gtf`: *Optional*, defaults to the latest _Salmon salar_ gtf which
is located at:
`/mnt/users/fabig/RNAseq/Ssa_genome/CIG_3.6v2_chrom-NCBI/GTF/Salmon_3p6_Chr_NCBI_230415.gtf`

- `-r|--read` *Optional*  Agrument for STAR if the read length is longer than
  2x250bp, specify `-r long`. If the read length is not longer than
  2x250bp this argument can be left empty. By default reads shorter
  than  2x250bp are assumed. *NOTE* this may change,
  depending on if the STAR version gets re-compiled...

---

`RNAmap` creates a folder tree in the current directory:

- `slurm` folder to collect the slurm reports. *Note* each slurm report lists:
   - date of job submission
   - version of the used module 
- `bash` folder to collect the bash scripts, generated and submitted through `RNAmap`.
- `fastq_trim` folder for the adaptor and quality trimmed .fastq files.
- `fastq_trim_pe` Same as above, read pairs where one 1 of the reads
  was found to be too short ( < 40bp ) where removed. Necessary for
  STAR to function properly. 
- `qc` folder for the quality control reports of the quality trimmed .fastq files.
- `star` folder contains read alignments to the genome in `.BAM` format. Reads were aligned using the program `STAR`
- `count` folder contains read counts per gene generated by the `HTSeq-count` script. 

## _master sheet_ format guide

Sample Nr.| Sample Name | TruSeq adaptor ID | TruSeq adaptor seq | FileBase |
---------|-----------|---------------|----------------|--------|
1	| 0_1 |	A001   | ATCACG |	0-1_S1 |
2	CTR_1 | A002 | CGATGT	| CTR-1_S2 |
3	MA_1  | A003 |  TTAGGC |	MA-1_S3 |
