# RNA_map

`RNA_map` is a bash script that automates the processes of:

1. Trimming of adapter sequences and poor quality nucleotides
   (cutadapt); Removing read pairs where one of the reads has less than 40bp 
2. Quality Control (FastQC)
3. Mapping reads to the genome (STAR)
4. Read counting per gene (HTSeq)
5. Summary statistics
6. Copy to common folder

The script currently works only if the following conditions are met
(possible to extend the script for other settings):
- Libraries were prepared using a *TruSeq stranded Kit*
- fastq data was generated on a MiSeq/NextSeq/HiSeq

## Example

It's usually a *good idea* to run the script with the option: `--execute
no`; Check that the *number of samples* as well as the ID of the first
and last sample are correct; and than run the whole stuff. 

```bash
script=/mnt/users/fabig/cluster_pipelines/RnaMapping/RNA_map.sh
fdir=/mnt/SeqData2/MiSeq_data/150612_M02210_0008_000000000-ACGHJ/Data/Intensities/BaseCalls
sdir=/mnt/SeqData2/MiSeq_data/150612_M02210_0008_000000000-ACGHJ/Data/Intensities/BaseCalls/SampleSheet.csv

bash $script -d $fdir -s $sdir -m Samples.txt -r long --copy test --desc description.txt --execute no
```

### Reports

The script also submits 2 `R` scripts that generate both:

1. STAR mapping statistics (see: `star/STAR_Log_Stat.txt` and `star/STAR_Log_Stat.pdf`)
2. HTseq counting statistics (see: `count/HTSeq_Count_Stats.txt` and `count/HTSeq_Count_Stats.pdf`)

In order for the `R` scripts to work you need to have the following
packages installed in your `orion R`

- `ggplot2`
- `reshape`
- `gtable`

## RNA_map options

- `-d|--dirin`: *Required* Full path of the directory that holds the .fasq.gz
  files

- `-g|--genome`: *Optional*, defaults to the latest version of the
  _Salmon salar_ genome,  located at:
`/mnt/users/fabig/RNAseq/Ssa_genome/CIG_3.6v2_chrom-NCBI/STAR_index`

- `--gtf`: *Optional*, defaults to the latest _Salmon salar_ gtf which
is located at:
`/mnt/users/fabig/Ssa_genome/CIG_3.6v2_chrom-NCBI/GTF/Salmon_3p6_Chr_070715_All.filter.gtf`

- `-m|--mastersheet`: *Required*  needs to point to a _master sheet_
  file (see [ _master sheet_ format guide])

- `-r|--read`: *Optional*  `<short/long>` Options passed on to STAR.
  Use <long> if the reads are longer than 2x250bp. Defaults to `<short>`.

- `--stranded`:*Optional*  `<yes/no/reverse>` Options passed on to
  `htseq-count` script. Defaults to `<reverse>` (for illumina stranded
  data); see
  [HtSeq](http://www-huber.embl.de/users/anders/HTSeq/doc/count.html).

- `--trimmer`: *Optional*  `<cutadapt/trimmomatic>`; Defaults to `<cutadapt>`

- `--copy`: *Required* Name of the common folder where a copy of
  - count
  - mapping_summary
  - mastersheet
  will be placed. When set to `<no>`, nothing will be copied

- `--desc`: *Required* Short description file giving the basic informations about
  the experiment ( see [format guide](## _description file_ format guid)). **NOTE** If you
  use the `--copy no` then you do *NOT* need to provide file here. 

- `--execute`: If set to `<no>` all folders/scripts will be genearted,
  but none of the jobs will be executed. 

---

`RNA_map` creates a folder tree in the current directory:

- `slurm` folder to collect the slurm reports. *Note* each slurm report lists:
   - date of job submission
   - version of the used module 
- `bash` folder to collect the bash scripts, generated and submitted through `RNA_map`.
- `fastq_trim` folder for the adaptor and quality trimmed .fastq
  file, read pairs where  one 1 of the reads
  was found to be too short ( < 40bp ) where removed. Necessary for
  STAR to function properly. 
- `qc` folder for the quality control reports of the quality trimmed .fastq files.
- `star` folder contains read alignments to the genome in `.BAM` format. Reads were aligned using the program `STAR`
- `count` folder contains read counts per gene generated by the `HTSeq-count` script. 
- `mapp_summary` folder that contains summarie statistics from STAR
  and HTSeq.

## _master sheet_ format guide

Plain tab delimited text file, no quotes, no header. *Important*
writing your _master sheet_ in excel works.

Given 3 samples:

- Sample 1: Adaptor A001 (ATCACG)
	- `0-1_S1_L001_R1_001.fastq.gz`
	- `0-1_S1_L001_R2_001.fastq.gz`
- Sample 2: Adaptor A002 (CGATGT)
	- `CTR-1_S2_L001_R1_001.fastq.gz`
	- `CTR-1_S2_L00_R2_001.fastq.gz`
- Sample 3: Adaptor A003 (TTAGGC)
	- `MA-1_S3_L001_R1_001.fastq.gz`
	- `MA-1_S3_L001_R2_001.fastq.gz`

The _master sheet_ should look like this (*WITH header*):

sample |	base |	adapter.id |	adapter.seq 
----|-----|-----|-----
0_1 |  0-1_S1_L001 |	A001   | ATCACG 
CTR_1 |  CTR-1_S2_L001 | A002 | CGATGT
MA_1  | MA-1_S3_L001|  A003 |  TTAGGC

## _description file_ format guide

The following information should be contained in the description file:

1. Organism
2. Tissue
3. Date, contact person
4. 2-3 Sentences about goal/setup of the experiment
